File: src/server/webfinger.js
const ar = require('../utils/ar')

module.exports = ({ domain, user }) => {
  const url = new URL(domain)
  const resource = `acct:${user}@${url.hostname}`

  const object = {
    subject: resource,
    links: [
      {
        rel: 'self',
        type: 'application/activity+json',
        href: new URL(domain).origin + '/'
      }
    ]
  }

  return ar(async (req, res) => {
    if (req.query.resource !== resource) {
      return res.sendStatus(404)
    }

    return res.json(object)
  })
}

File: eagle/webfinger.go
package eagle

type WebFinger struct {
	Subject string          `json:"subject"`
	Aliases []string        `json:"aliases,omitempty"`
	Links   []WebFingerLink `json:"links,omitempty"`
}

type WebFingerLink struct {
	Href string `json:"href"`
	Rel  string `json:"rel,omitempty"`
	Type string `json:"type,omitempty"`
}
